File: scripts/quick_stats.py
#!/usr/bin/env python3
"""
Quick statistics from cluster results CSV (no dependencies beyond pandas).
Prints summary to terminal and saves to text file.

Usage: python3 quick_stats.py cluster_results.csv [output.txt]
"""

import ast
import pandas as pd
import sys
from pathlib import Path


def load_and_clean(csv_path):
    df = pd.read_csv(csv_path)
    bool_cols = ['is_sat', 'is_unsat', 'has_proof', 'proof_truncated', 'has_error']
    for col in bool_cols:
        if col in df.columns:
            df[col] = df[col].map({'true': True, 'false': False, True: True, False: False})
    return df


def parse_list_col(series):
    """Parse a column of stringified lists like '[50,42,39]' into a list of lists."""
    result = []
    for v in series.dropna():
        try:
            parsed = ast.literal_eval(v) if isinstance(v, str) else v
            if isinstance(parsed, list):
                result.append(parsed)
        except Exception:
            pass
    return result


def print_section(title):
    print(f"\n{'='*70}")
    print(f" {title}")
    print('='*70)


def main():
    if len(sys.argv) < 2:
        print("Usage: python3 quick_stats.py cluster_results.csv [output.txt]")
        sys.exit(1)

    csv_file = sys.argv[1]
    output_file = sys.argv[2] if len(sys.argv) > 2 else "stats_summary.txt"

    if not Path(csv_file).exists():
        print(f"Error: {csv_file} not found")
        sys.exit(1)

    print(f"Loading {csv_file}...")
    df = load_and_clean(csv_file)

    class Tee:
        def __init__(self, *files):
            self.files = files
        def write(self, data):
            for f in self.files:
                f.write(data)
        def flush(self):
            for f in self.files:
                f.flush()

    with open(output_file, 'w') as f:
        original_stdout = sys.stdout
        sys.stdout = Tee(sys.stdout, f)

        print(f"Cluster Results Analysis")
        print(f"CSV: {csv_file}")
        print(f"=" * 70)

        # ── Overview ──────────────────────────────────────────────────────────
        print_section("OVERVIEW")
        n = len(df)
        print(f"Total instances:      {n:>10,}")
        for col, label in [('is_sat','SAT'), ('is_unsat','UNSAT'), ('has_proof','With proof'),
                           ('proof_truncated','Truncated proofs'), ('has_error','Errors')]:
            if col in df.columns:
                c = df[col].sum()
                print(f"{label+':':<22} {int(c):>10,}  ({c/n*100:.1f}%)")

        if 'resolv_iterations' in df.columns:
            rc = (df['resolv_iterations'] > 0).sum()
            print(f"{'Resolv ran:':<22} {rc:>10,}  ({rc/n*100:.1f}%)")

        # ── Skip / error breakdown ─────────────────────────────────────────────
        if 'skip_reason' in df.columns:
            skip_counts = df[df['skip_reason'].notna()]['skip_reason'].value_counts()
            if not skip_counts.empty:
                print_section("SKIP REASONS")
                for reason, count in skip_counts.items():
                    print(f"  {reason:<35} {count:>7,}  ({count/n*100:.1f}%)")

        if 'error_type' in df.columns and df['has_error'].sum() > 0:
            error_counts = df[df['has_error'] == True]['error_type'].value_counts()
            if not error_counts.empty:
                print_section("ERROR TYPES")
                for err_type, count in error_counts.items():
                    print(f"  {err_type:<35} {count:>7,}  ({count/df['has_error'].sum()*100:.1f}%)")

        # ── Instances with proofs ──────────────────────────────────────────────
        proof_df = df[df['has_proof'] == True].copy()

        if proof_df.empty:
            print("\nNo instances with proofs found.")
            sys.stdout = original_stdout
            return

        # ── Timing ────────────────────────────────────────────────────────────
        print_section(f"TIMING STATISTICS  (n={len(proof_df):,} instances with proofs)")
        for label, col in [('Parse Time','grim_parse_time'), ('Trim Time','grim_trim_time'),
                           ('Write Time','grim_write_time'), ('Total Time','grim_total_time')]:
            if col in proof_df.columns:
                data = proof_df[col].dropna()
                if data.empty:
                    continue
                print(f"\n  {label}:")
                print(f"    Mean      {data.mean():>9.2f} s   Median {data.median():>9.2f} s")
                print(f"    Min       {data.min():>9.2f} s   Max    {data.max():>9.2f} s")
                print(f"    95th %ile {data.quantile(0.95):>9.2f} s   Std    {data.std():>9.2f} s")

        # ── Size / reduction ──────────────────────────────────────────────────
        print_section(f"SIZE & REDUCTION  (n={len(proof_df):,})")

        if 'inp_total_size' in proof_df.columns and 'grim_total_size' in proof_df.columns:
            valid = proof_df[['inp_total_size','grim_total_size']].dropna()
            if not valid.empty:
                inp_gb = valid['inp_total_size'].sum() / 1024**3
                out_gb = valid['grim_total_size'].sum() / 1024**3
                print(f"  Total input  {inp_gb:>8.2f} GB   Total output {out_gb:>8.2f} GB   "
                      f"Reduction {(1-out_gb/inp_gb)*100:.1f}%")

        if 'inp_total_nbeq' in proof_df.columns and 'grim_total_cone' in proof_df.columns:
            valid = proof_df[['inp_total_nbeq','grim_total_cone']].dropna()
            valid = valid[valid['inp_total_nbeq'] > 0]
            if not valid.empty:
                ratio = (valid['inp_total_nbeq'] - valid['grim_total_cone']) / valid['inp_total_nbeq']
                proof_df.loc[valid.index, 'constraint_reduction'] = ratio
                print(f"\n  Constraint reduction  mean {ratio.mean():>6.1%}   "
                      f"median {ratio.median():>6.1%}   "
                      f"min {ratio.min():>6.1%}   max {ratio.max():>6.1%}")

        if 'grim_cone_literals' in proof_df.columns and 'grim_smol_literals' in proof_df.columns:
            valid = proof_df[['grim_cone_literals','grim_smol_literals']].dropna()
            valid = valid[valid['grim_cone_literals'] > 0]
            if not valid.empty:
                ratio = (valid['grim_cone_literals'] - valid['grim_smol_literals']) / valid['grim_cone_literals']
                proof_df.loc[valid.index, 'literal_reduction'] = ratio
                print(f"  Literal reduction     mean {ratio.mean():>6.1%}   "
                      f"median {ratio.median():>6.1%}   "
                      f"min {ratio.min():>6.1%}   max {ratio.max():>6.1%}")

        # ── Cone step types (M2) ───────────────────────────────────────────────
        rup_col = 'grim_cone_rup'
        if rup_col in proof_df.columns and proof_df[rup_col].notna().any():
            step_cols = ['grim_cone_rup','grim_cone_pol','grim_cone_red','grim_cone_ia']
            step_df = proof_df[[c for c in step_cols if c in proof_df.columns]].dropna()
            if not step_df.empty:
                print_section(f"CONE STEP TYPES  (n={len(step_df):,})")
                total_steps = step_df.sum(axis=1)
                for col in step_df.columns:
                    label = col.replace('grim_cone_','').upper()
                    mean_n   = step_df[col].mean()
                    mean_pct = (step_df[col] / total_steps.replace(0, float('nan'))).mean()
                    print(f"  {label:<6}  mean count {mean_n:>9.1f}   mean share {mean_pct:>6.1%}   "
                          f"median {step_df[col].median():>7.0f}   max {step_df[col].max():>7.0f}")

        # ── Cone depth (M2) ───────────────────────────────────────────────────
        if 'grim_cone_depth_max' in proof_df.columns and proof_df['grim_cone_depth_max'].notna().any():
            dmax = proof_df['grim_cone_depth_max'].dropna()
            dmean = proof_df['grim_cone_depth_mean'].dropna() if 'grim_cone_depth_mean' in proof_df.columns else None
            print_section(f"CONE DAG DEPTH  (n={len(dmax):,})")
            print(f"  Max depth   mean {dmax.mean():>7.1f}   median {dmax.median():>7.1f}   "
                  f"min {dmax.min():>5.0f}   max {dmax.max():>5.0f}")
            if dmean is not None and not dmean.empty:
                print(f"  Mean depth  mean {dmean.mean():>7.2f}   median {dmean.median():>7.2f}   "
                      f"min {dmean.min():>5.2f}   max {dmean.max():>5.2f}")

        # ── Resolv loop (M2) ──────────────────────────────────────────────────
        if 'resolv_stop_reason' in df.columns and df['resolv_stop_reason'].notna().any():
            print_section("RESOLV LOOP")

            reasons = df['resolv_stop_reason'].value_counts()
            print(f"  Stop reason breakdown:")
            for reason, count in reasons.items():
                print(f"    {reason:<25} {count:>6,}  ({count/len(df)*100:.1f}%)")

            if 'resolv_iter_pat_nodes' in df.columns:
                lists = parse_list_col(df['resolv_iter_pat_nodes'])
                if lists:
                    n_iters = [len(l) - 1 for l in lists]   # iterations = len - 1 (iter 0 is initial)
                    shrinkages = [(l[0] - l[-1]) / l[0] for l in lists if l[0] > 0]
                    print(f"\n  Pattern graph shrinkage  (n={len(lists):,} instances that ran resolv):")
                    print(f"    Iterations  mean {sum(n_iters)/len(n_iters):.1f}   "
                          f"max {max(n_iters)}")
                    if shrinkages:
                        import statistics
                        print(f"    Node shrink mean {sum(shrinkages)/len(shrinkages):.1%}   "
                              f"median {statistics.median(shrinkages):.1%}   "
                              f"max {max(shrinkages):.1%}")

        # ── UNSAT core ────────────────────────────────────────────────────────
        if 'core_pattern_nodes' in proof_df.columns and proof_df['core_pattern_nodes'].notna().any():
            core_df = proof_df[proof_df['core_pattern_nodes'].notna()].copy()
            print_section(f"UNSAT CORE STATISTICS  (n={len(core_df):,})")
            for label, col in [('Pattern core nodes','core_pattern_nodes'),
                                ('Target core nodes', 'core_target_nodes')]:
                if col in core_df.columns:
                    data = core_df[col].dropna()
                    if not data.empty:
                        print(f"  {label:<25}  mean {data.mean():>8.1f}   "
                              f"median {data.median():>8.1f}   max {data.max():>8.0f}")

        # ── Outliers ──────────────────────────────────────────────────────────
        print_section("OUTLIERS (>1.5 IQR above Q3)")
        for label, col in [('Total Time','grim_total_time'), ('Trim Time','grim_trim_time')]:
            if col not in proof_df.columns:
                continue
            data = proof_df[col].dropna()
            Q1, Q3 = data.quantile(0.25), data.quantile(0.75)
            outliers = proof_df[proof_df[col] > Q3 + 1.5*(Q3-Q1)]
            if not outliers.empty:
                print(f"\n  {label} outliers ({len(outliers)}):")
                for _, row in outliers.nlargest(5, col).iterrows():
                    inp  = int(row['inp_total_nbeq'])  if 'inp_total_nbeq'  in row and pd.notna(row['inp_total_nbeq'])  else 0
                    cone = int(row['grim_total_cone']) if 'grim_total_cone' in row and pd.notna(row['grim_total_cone']) else 0
                    print(f"    {row['instance']:<35} {row[col]:>8.1f}s  inp={inp:>8,}  cone={cone:>7,}")

        # ── Top 10 slowest ────────────────────────────────────────────────────
        if 'grim_total_time' in proof_df.columns:
            print_section("TOP 10 SLOWEST")
            for _, row in proof_df.nlargest(10, 'grim_total_time').iterrows():
                inp  = int(row['inp_total_nbeq'])  if pd.notna(row.get('inp_total_nbeq',  float('nan'))) else 0
                cone = int(row['grim_total_cone']) if pd.notna(row.get('grim_total_cone', float('nan'))) else 0
                print(f"  {row['instance']:<35} {row['grim_total_time']:>8.1f}s  "
                      f"inp={inp:>8,}  cone={cone:>7,}")

        # ── Top 10 best constraint reductions ─────────────────────────────────
        if 'constraint_reduction' in proof_df.columns:
            print_section("TOP 10 BEST CONSTRAINT REDUCTIONS")
            for _, row in proof_df.nlargest(10, 'constraint_reduction').iterrows():
                inp  = int(row['inp_total_nbeq'])  if pd.notna(row.get('inp_total_nbeq',  float('nan'))) else 0
                cone = int(row['grim_total_cone']) if pd.notna(row.get('grim_total_cone', float('nan'))) else 0
                print(f"  {row['instance']:<35} {row['constraint_reduction']:>6.1%}  "
                      f"({inp:>8,} → {cone:>7,})")

        print("\n" + "="*70)
        print(f"Summary saved to: {output_file}")
        sys.stdout = original_stdout

    print(f"\nAnalysis complete. Results saved to: {output_file}")


if __name__ == '__main__':
    main()

File: scripts/quick_stats.jl
#!/usr/bin/env julia
# Quick terminal statistics from cluster_results.csv — stdlib only, no package deps.
# Usage: julia scripts/quick_stats.jl cluster_results.csv [output.txt]

using Statistics, Printf

struct TeeIO <: IO; a::IO; b::IO; end
Base.write(t::TeeIO, x::UInt8) = (write(t.a, x); write(t.b, x); 1)

sec(io, title) = println(io, "\n" * "="^70 * "\n " * title * "\n" * "="^70)
pct(n, tot)    = tot > 0 ? @sprintf("%.1f%%", 100n / tot) : "—"

function smry(data, unit="")
    isempty(data) && return "(no data)"
    @sprintf("mean %8.2f%s   median %8.2f%s   min %6.2f%s   max %6.2f%s",
        mean(data), unit, median(data), unit, minimum(data), unit, maximum(data), unit)
end

function countmap(v)
    d = Dict{eltype(v), Int}()
    for x in v; d[x] = get(d, x, 0) + 1; end
    d
end

# ── CSV parser (handles quoted fields containing commas) ──────────────────────

function csv_fields(line)
    fields = String[]; i = 1; n = length(line)
    while i <= n
        if i <= n && line[i] == '"'
            j = i + 1
            while j <= n && line[j] != '"'; j += 1; end
            push!(fields, line[i+1:j-1])
            i = j + 2
        else
            j = findnext(==(','), line, i)
            if j === nothing
                push!(fields, line[i:end]); break
            else
                push!(fields, line[i:j-1]); i = j + 1
            end
        end
    end
    fields
end

function read_csv(path)
    lines = readlines(path)
    isempty(lines) && error("empty CSV")
    header = [strip(h, '"') for h in split(lines[1], ',')]
    cols   = Dict(h => String[] for h in header)
    for line in @view lines[2:end]
        isempty(strip(line)) && continue
        fields = csv_fields(line)
        for (i, h) in enumerate(header)
            push!(cols[h], i <= length(fields) ? fields[i] : "")
        end
    end
    cols, max(0, length(lines) - 1)
end

numcol(cols, name) = Float64[x for x in
    (tryparse(Float64, v) for v in get(cols, name, String[])) if x !== nothing]

boolcol(cols, name) = [v == "true" for v in get(cols, name, String[])]

# ── Main ─────────────────────────────────────────────────────────────────────

function main()
    length(ARGS) < 1 && (println("Usage: julia quick_stats.jl <csv> [output.txt]"); exit(1))
    csv_file = ARGS[1]; out_file = length(ARGS) >= 2 ? ARGS[2] : "stats_summary.txt"
    isfile(csv_file) || (println("Error: $csv_file not found"); exit(1))

    println("Loading $csv_file...")
    cols, n = read_csv(csv_file)

    has_proof = boolcol(cols, "has_proof")
    # Glasgow writes status = true (SAT) / false (UNSAT); derive from status column
    status_col = get(cols, "status", String[])
    is_sat    = !isempty(status_col) ? [v == "true"  for v in status_col] : boolcol(cols, "is_sat")
    is_unsat  = !isempty(status_col) ? [v == "false" for v in status_col] : boolcol(cols, "is_unsat")
    has_error = boolcol(cols, "has_error")
    truncated = boolcol(cols, "proof_truncated")
    resolv_it = numcol(cols, "resolv_iterations")

    # proof-only row mask
    proof_idx = findall(has_proof)
    pcol(name) = numcol(Dict(name => get(cols, name, String[])[proof_idx]), name)

    open(out_file, "w") do f
        io = TeeIO(stdout, f)

        println(io, "Cluster Results Analysis — $csv_file")
        println(io, "="^70)

        # ── Overview ──────────────────────────────────────────────────────────
        sec(io, "OVERVIEW")
        println(io, @sprintf("Total instances:      %10d", n))
        for (v, label) in [(is_sat,"SAT"), (is_unsat,"UNSAT"), (has_proof,"With proof"),
                           (truncated,"Truncated"), (has_error,"Errors")]
            c = sum(v)
            println(io, @sprintf("%-22s %10d  (%s)", label * ":", c, pct(c, n)))
        end
        if !isempty(resolv_it)
            rc = sum(resolv_it .> 0)
            ri = sum(.!isempty.(get(cols, "resolv_stop_reason", String[])))  # invoked (incl. 0-iter)
            println(io, @sprintf("%-22s %10d  (%s)", "Resolv invoked:", ri, pct(ri, n)))
            println(io, @sprintf("%-22s %10d  (%s of invoked)", "Resolv ≥1 iter:", rc, pct(rc, ri)))
        end

        np = length(proof_idx)
        np == 0 && (println(io, "\nNo instances with proofs."); return)

        # ── Skip / error breakdown ────────────────────────────────────────────
        skip_vals = get(cols, "skip_reason", String[])[proof_idx]
        skip_vals = filter(!isempty, skip_vals)
        if !isempty(skip_vals)
            sec(io, "SKIP REASONS")
            for (reason, count) in sort(collect(countmap(skip_vals)); by=x -> -x[2])
                println(io, @sprintf("  %-35s %7d  (%s)", reason, count, pct(count, n)))
            end
        end

        # ── Timing ────────────────────────────────────────────────────────────
        sec(io, "TIMING  (n=$np instances with proofs)")
        for (label, col) in [("Parse Time","grim_parse_time"), ("Trim Time","grim_trim_time"),
                              ("Write Time","grim_write_time"), ("Total Time","grim_total_time")]
            data = pcol(col); isempty(data) && continue
            println(io, "\n  $label:")
            println(io, "    " * smry(data, "s"))
            println(io, @sprintf("    95th %%ile %8.2fs   std %8.2fs", quantile(data, 0.95), std(data)))
        end

        # ── Size / reduction ──────────────────────────────────────────────────
        sec(io, "SIZE & REDUCTION  (n=$np)")
        inp_sz = pcol("inp_total_size"); out_sz = pcol("grim_total_size")
        if !isempty(inp_sz) && !isempty(out_sz)
            println(io, @sprintf("  Total input %8.2f GB   Total output %8.2f GB   Reduction %.1f%%",
                sum(inp_sz)/1024^3, sum(out_sz)/1024^3, (1 - sum(out_sz)/sum(inp_sz))*100))
        end
        inp_eq = pcol("inp_total_nbeq"); out_eq = pcol("grim_total_cone")
        if !isempty(inp_eq) && !isempty(out_eq)
            r = [(a-b)/a for (a,b) in zip(inp_eq, out_eq) if a > 0]
            isempty(r) || println(io, @sprintf(
                "  Constraint reduction  mean %.1f%%   median %.1f%%   min %.1f%%   max %.1f%%",
                mean(r)*100, median(r)*100, minimum(r)*100, maximum(r)*100))
        end
        lit_in = pcol("grim_cone_literals"); lit_out = pcol("grim_smol_literals")
        if !isempty(lit_in) && !isempty(lit_out)
            r = [(a-b)/a for (a,b) in zip(lit_in, lit_out) if a > 0]
            isempty(r) || println(io, @sprintf(
                "  Literal reduction     mean %.1f%%   median %.1f%%   min %.1f%%   max %.1f%%",
                mean(r)*100, median(r)*100, minimum(r)*100, maximum(r)*100))
        end

        # ── Cone step types ───────────────────────────────────────────────────
        rup_data = pcol("grim_cone_rup")
        if !isempty(rup_data)
            sec(io, "CONE STEP TYPES  (n=$(length(rup_data)))")
            for (col, label) in [("grim_opb_cone","OPB"), ("grim_cone_rup","RUP"),
                                  ("grim_cone_pol","POL"), ("grim_cone_ia","IA"),
                                  ("grim_cone_red","RED")]
                data = pcol(col); isempty(data) && continue
                println(io, @sprintf("  %-6s  mean %9.1f   median %7.0f   max %7.0f",
                    label, mean(data), median(data), maximum(data)))
            end
            total_v = pcol("grim_total_cone")
            println(io, "\n  Fractions of total cone (OPB + derived):")
            for (col, label) in [("grim_opb_cone","OPB"), ("grim_cone_rup","RUP"),
                                  ("grim_cone_pol","POL"), ("grim_cone_ia","IA"),
                                  ("grim_cone_red","RED")]
                num_v = pcol(col); (isempty(num_v) || isempty(total_v)) && continue
                fracs = [n/t for (n,t) in zip(num_v, total_v) if t > 0]
                isempty(fracs) && continue
                println(io, @sprintf("  %-6s  mean %6.1f%%   median %6.1f%%",
                    label, mean(fracs)*100, median(fracs)*100))
            end
        end

        # ── Cone depth ────────────────────────────────────────────────────────
        dmax = pcol("grim_cone_depth_max")
        if !isempty(dmax)
            sec(io, "CONE DAG DEPTH  (n=$(length(dmax)))")
            println(io, "  Max depth   " * smry(dmax))
            dmean = pcol("grim_cone_depth_mean")
            isempty(dmean) || println(io, "  Mean depth  " * smry(dmean))
            for (label, col, note) in [
                    ("P50        ", "grim_cone_depth_p50", ""),
                    ("P90        ", "grim_cone_depth_p90", ""),
                    ("Entropy    ", "grim_cone_depth_entropy", "  (0=chain, high=spread)"),
                    ("Bottom frac", "grim_cone_bottom_frac",  "  (fraction of steps at depth ≤ 2)"),
                    ("Width CV   ", "grim_cone_width_cv",     "  (0=uniform, high=spiky)")]
                data = pcol(col); isempty(data) && continue
                println(io, @sprintf("  %s  mean %8.3f   median %8.3f%s",
                    label, mean(data), median(data), note))
            end
            println(io)
            for (label, col, note) in [
                    ("POL depth mean", "grim_cone_pol_depth_mean", "centroid in depth space"),
                    ("POL ante mean ", "grim_cone_pol_ante_mean",  "avg antecedents per POL step"),
                    ("POL OPB frac  ", "grim_cone_pol_opb_frac",   "fraction of POL antes from axioms")]
                data = pcol(col); isempty(data) && continue
                println(io, @sprintf("  %s  mean %6.2f   (%s)", label, mean(data), note))
            end
        end

        # ── Resolv ────────────────────────────────────────────────────────────
        reasons = filter(!isempty, get(cols, "resolv_stop_reason", String[]))
        if !isempty(reasons)
            sec(io, "RESOLV LOOP  (n=$(length(reasons)) invoked)")
            for (reason, count) in sort(collect(countmap(reasons)); by=x -> -x[2])
                println(io, @sprintf("  %-30s %6d  (%s of invoked)", reason, count, pct(count, length(reasons))))
            end
            for (label, col) in [("Pattern","resolv_pat_shrinkage"), ("Target","resolv_tar_shrinkage")]
                ps = filter(x -> x > 0, numcol(cols, col))
                isempty(ps) && continue
                println(io, @sprintf("\n  %s shrinkage (n=%d where > 0):  mean %.1f%%   median %.1f%%   max %.1f%%",
                    label, length(ps), mean(ps)*100, median(ps)*100, maximum(ps)*100))
            end
        end

        # ── UNSAT core ────────────────────────────────────────────────────────
        cpn = pcol("core_pattern_nodes")
        if !isempty(cpn)
            sec(io, "UNSAT CORE  (n=$(length(cpn)))")
            println(io, "  Pattern core nodes  " * smry(cpn))
            ctn = pcol("core_target_nodes")
            isempty(ctn) || println(io, "  Target  core nodes  " * smry(ctn))
        end

        # ── Outliers ──────────────────────────────────────────────────────────
        sec(io, "OUTLIERS (>1.5 IQR above Q3)")
        for (label, col) in [("Total Time","grim_total_time"), ("Trim Time","grim_trim_time")]
            data = pcol(col); isempty(data) && continue
            q1, q3 = quantile(data, 0.25), quantile(data, 0.75)
            cutoff = q3 + 1.5*(q3 - q1)
            out_idx = [proof_idx[i] for (i,v) in enumerate(data) if v > cutoff]
            isempty(out_idx) && continue
            println(io, "\n  $label outliers ($(length(out_idx))):")
            order = sortperm([parse(Float64, get(cols, col, String[])[i]) for i in out_idx]; rev=true)
            for i in out_idx[order[1:min(5,end)]]
                t    = parse(Float64, cols[col][i])
                inp  = something(tryparse(Int, cols["inp_total_nbeq"][i]),  0)
                cone = something(tryparse(Int, cols["grim_total_cone"][i]), 0)
                println(io, @sprintf("    %-35s %8.1fs  inp=%8d  cone=%7d",
                    cols["instance"][i], t, inp, cone))
            end
        end

        # ── Top 10 slowest ────────────────────────────────────────────────────
        times = pcol("grim_total_time")
        if !isempty(times)
            sec(io, "TOP 10 SLOWEST")
            order = sortperm(times; rev=true)
            for i in order[1:min(10,end)]
                gi = proof_idx[i]
                inp  = something(tryparse(Int, cols["inp_total_nbeq"][gi]),  0)
                cone = something(tryparse(Int, cols["grim_total_cone"][gi]), 0)
                println(io, @sprintf("  %-35s %8.1fs  inp=%8d  cone=%7d",
                    cols["instance"][gi], times[i], inp, cone))
            end
        end

        println(io, "\n" * "="^70)
        println(io, "Summary saved to: $out_file")
    end
    println("\nDone.")
end

main()
